### PR TITLE
Adding initial API Product revision implementation

### PR DESCRIPTION
**Pull request by CrowleyRajapakse**

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 1]
```diff
@@ -8195,10 +8212,22 @@ public void saveToGateway(APIProduct product) throws APIManagementException {
         List<APIProductResource> productResources = product.getProductResources();
 
         //Only publish to gateways if the state is in Published state and has atleast one resource
-        if("PUBLISHED".equals(product.getState()) && !productResources.isEmpty()) {
-            publishToGateway(product);
-        }
-
+//        if("PUBLISHED".equals(product.getState()) && !productResources.isEmpty()) {
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java:8215]: lets remove the commented code

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 2]
```diff
@@ -8221,12 +8250,13 @@ public void deleteAPIProduct(APIProduct apiProduct) throws APIManagementExceptio
 
             // gatewayType check is required when API Management is deployed on
             // other servers to avoid synapse
-            if (gatewayExists && "Synapse".equals(gatewayType)) {
-                removeFromGateway(apiProduct, tenantDomain);
-
-            } else {
-                log.debug("Gateway is not existed for the current API Provider");
-            }
+//            if (gatewayExists && "Synapse".equals(gatewayType)) {
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java:8253]: remove

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 3]
```diff
@@ -8324,11 +8354,18 @@ public void deleteAPIProduct(APIProductIdentifier identifier, String apiProductU
                 }
             }
         }
-
-
-        if (resources.size() > 0) {
-            publishToGateway(product);
-        }
+//        Map<String, Map<String, String>> failedGateways = new ConcurrentHashMap<String, Map<String, String>>();
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java:8357]: remove

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java (hunk 1 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java (hunk 2 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java (hunk 3 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: resolved

**Pull request merged**